### ISSUE TITLE
Context-free grammar inference

### ISSUE THREAD

**Issue by gsvgit** [YaccConstructor/YaccConstructor#213]
Implement CF grammar inference for biological sequences (16s, etc).
Output format -- YARD (IL) grammar specification language.

- [ ] Theoretical comparison of CF grammar inference approaches and algorithms.
- [ ] Implementation of selected algorithm(s)
- [ ] Evaluation on 16s datasets

Some materials for start:
- [Axiomatization of Finite Algebras](https://arxiv.org/abs/1403.7347)
- (!!!)[A nondeterministic approach to infer context free grammar from sequence](http://ieeexplore.ieee.org/abstract/document/7073350/)
- [Learning Conjunctive Grammars and Contextual Binary Feature Grammars](http://link.springer.com/chapter/10.1007%2F978-3-319-15579-1_49#page-1)
- [Context free grammar induction using genetic algorithms](http://ieeexplore.ieee.org/abstract/document/243118/)
- [Learning Trees from Strings: A Strong Learning Algorithm for some Context-Free Grammars](http://www.jmlr.org/papers/volume14/clark13a/clark13a.pdf)
- [Searching for smallest grammars on large sequences and application to DNA](http://www.sciencedirect.com/science/article/pii/S1570866711000517)
- [Learning Recursive Automata from Positive Examples](https://hal.inria.fr/inria-00470101/document/)
---
- [Grammar-based classifier system](http://dl.acm.org/citation.cfm?id=1486700)
- [Learning context-free grammar using improved tabular representation](https://www.researchgate.net/publication/222231252_Learning_context-free_grammar_using_improved_tabular_representation)
---
- [General Perspective on Distributionally Learnable Classes. Ryo Yoshinaka](http://www.anthology.aclweb.org/W/W15/W15-23.pdf#page=97)
- [The Strong, Weak, and Very Weak Finite Context and Kernel Properties](http://research.nii.ac.jp/~kanazawa/publications/swvwfckp.pdf)
- [E-Generalization Using Grammars. Jochen Burghardt](https://arxiv.org/pdf/1403.8118v1.pdf)

**Comment by gsvgit** [comment]
@antkhorin
First part.
- [ ] Fork -> Clone -> build YaccConstructor. Branch: GLL-FSA
- [ ] Create general grammar for stems with gaps (example of 16s grammar --  YaccConstructor\src\YC.GrammarZOO\Bio\16s\R16S_19_27.yrd)
- [ ] Try to parse 16s sequences (YaccConstructor\Tests\data\bio\16s\ ) with GLL. Examples of parsing with GLL: YaccConstructor\Tests\GLLParser.Simple.Tests\GLLParserSimpleTest.fs

**Comment by antkhorin** [comment]
- [x] Fork -> Clone -> build YaccConstructor. Branch: GLL-FSA
Здравствуйте.
Я попробовал распарсить 16s последовательности используя грамматики из файлов R16S_19_27.yrd и R16S_tails.yrd, но большинство последовательностей из файла HOMD_16S_rRNA_RefSeq_V14.5.fasta не подходят под грамматику, это нормально? И не совсем ясно что означает Create general grammar for stems with gaps. Я пробовал добавлять в грамматику эпсилон правила, но при любых изменениях грамматики разбор не заканчивался и Out of memory exception.

**Comment by gsvgit** [comment]
Здравствуйте.
Про *_tails.yrd не удивительно. Там странная грамматика, а вот R16S_19_27.yrd должна была, по моим представлениям, работать. Ну ладно. 

А что нужно: написать грамматику, в которой есть всевозможные комбинации стемов и несворачиваемых кусков. Должно получиться что-то типа сбалансированной скобочной последовательности с вставками несбалансированных кусков. Можно сразу ввести ограничения. Скажем, высота стема не меньше 2 пар, длина несворачиваемого участка не больше 20. Это всё в грамматике выразимо.

**Comment by antkhorin** [comment]
Ну вот я сделал грамматику [Grammar.txt](https://github.com/YaccConstructor/YaccConstructor/files/988392/Grammar.txt)
Так нужно было?

**Comment by antkhorin** [comment]
И вопрос про извлечение грамматики и задачу в общем.
Мне нужно будет сделать так: берем одну строку и строим настолько короткую грамматику для данной строки насколько получится, или так: берем множество весьма похожих строк (как в датасете) и строим одну грамматику, чтобы под нее подходило как можно больше строк из датасета и не подходили сильно отличающиеся от строк датасета.

**Comment by gsvgit** [comment]
Да, похоже на правду. Есть всякие детали, но они, вероятно, решатся в ходе
опытов. Например, я бы оставил только stem-ы (без gstem), так как в gstem
стягиваются G и A, что, вообще говоря, редкость.

По поводу задачи вообще. Надо построить грамматику по сравнительно
небольшому датасету, чтобы под неё подходили почти все цепочки из любого
большого датасета (да, у меня гиппотеза, что вторичная структура должна
быть очень близка). Но при этом грамматика не должна быть прямо таки
минимальной. То, что Вы сейчас написали -- это самая общая. Теперь её надо
сузить. По идее, в этом поможет sppf. Сколько времени занимает разбора,
скажем, 10 цепочек по этой грамматике с построением sppf?

Regards,
Semyon Grigorev.

10 мая 2017 г., 5:45 пользователь Anton Khorin <notifications@github.com>
написал:

**Comment by antkhorin** [comment]
В этой грамматике слишком много неоднозначности. Даже если оставить в стемах только связи A U и C G и ограничить длину несворачиваемой цепочки десятью, я не дождался окончания разбора (>15 минут).

**Comment by gsvgit** [comment]
@antkhorin 
1) Медленно работает -- не страшно. Данный шаг будет применяться примерно как обучение в машинном обучении. Один раз запустили, может хоть сутки считать, если потом использование результата позволит решать задачу быстро. Как именно -- по SPPF для некоторого набора цепочек нужно выбрать выбрать наилучшее приближение сверху, при этом такое, в котором больше всего "высоких" стемов. Кажется, это будет неплохое загрубление. Далее по этому дереву (вероятно, не совсем дереву) можно обратно построить уже более сложную, но более однозначную грамматику. парсится будет быстрее, при этом находить будет почти только нужные нам куски.

2) Можно для начала сократить цепочки. Брать из них подцепочки с 500 по 900 символы. И тут уже неплохо бы дождаться завершения и оценить время.

3) Нужно подумать, как из дерева построить грамматику.

**Comment by antkhorin** [comment]
Здравствуйте.
Мне по прежнему совершенно не ясно, что нужно делать. Я пробовал запускать YaccConstructor на различных тестах с разными модификациями грамматики. Все работает настолько медленно. что никаких результатов получить не удалось. Можно ли сделать так, чтобы при удачном разборе программа останавливалась, а не продолжался разбор всех возможных неоднозначностей (их получается невероятно много)? Достаточно ли мне просто реализовать алгоритм из статьи? Не могли бы вы сформулировать задание поточнее, чтобы я мог уложиться в сроки. Мне бы очень хотелось закрыть практику в университете до 25 июня.

**Comment by gsvgit** [comment]
1) 15 минут -- это ни о чём. Как я говорил, на построение грамматики можно
потратить часы и даже сутки.
2) Остановить нельзя. Нужен не какой-то один вариант, а оптимальный из всех
возможных. Так что построить придётся все.
3) Чтобы стало полегче, можете взять не целиком 16s, а подцепочку. скажем,
с 560 по 920 символ.
4) Я пока не видел статей с тем алгоритмом, который мы сейчас пытаемся
придумать.
5) Цель -- нужен механизм построения граммтик при наличии известных
цепочек, принадлежащих языку и некоторых знаниях о шаблонных конструкциях
(стемы и линейные участки).
Задачи:
-- Написать грамматику G -- наиболее общую (то, что Вы как-то сделали)
-- Осознать, что лес вывода (SPPF) -- это грамматика пересечения двух
языков: заданного грамматикой и заданного входной строкой. Но в этой
грамматике много лишнего: те самые неоднозначности. Надо вычленить
грамматику однозначную и соответствующую наиболее "плотной" свёртке -- это
одно из деревьев вывода. Наиболее "плотное". Сгенерировать эту грамматику в
Yard.
-- Попробовать по этой грамматике распарсить другие цепочки 16s. Скорее
всего не получится. Выяснить, почему.
-- Оформить всё, что выше в виде воспроизводимого эксперимента: код,
документация, примеры и т.д.


Regards,
Semyon Grigorev.

14 июня 2017 г., 5:55 пользователь Anton Khorin <notifications@github.com>
написал:
